executor/adapter.go

```diff
@@ -561,18 +565,121 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 	}, nil
 }
 
-func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor) error {
+func (a *ExecStmt) handleStmtForeignKeyTrigger(ctx context.Context, fkCtx FKCascadeContext, e Executor, isPessimistic bool) error {
+	if fkCtx.hasCascades {
+		// If the ExecStmt has foreign key cascade to be executed, we need call `StmtCommit` to commit the ExecStmt itself
+		// change first.
+		// Since `UnionScanExec` use `SnapshotIter` and `SnapshotGetter` to read txn mem-buffer, if we don't  do `StmtCommit`,
+		// then the fk cascade executor can't read the mem-buffer changed by the ExecStmt.
+		a.Ctx.StmtCommit()
+	}
+	err := a.handleForeignKeyTrigger(ctx, e, isPessimistic, 1)
+	if err != nil {
+		err1 := a.handleFKTriggerError(fkCtx)
+		if err1 != nil {
+			return errors.Errorf("handle foreign key trigger error failed, err: %v, original_err: %v", err1, err)
+		}
+		return err
+	}
+	if fkCtx.savepointName != "" {
+		a.Ctx.GetSessionVars().TxnCtx.ReleaseSavepoint(fkCtx.savepointName)
+	}
+	return nil
+}
+
+var maxForeignKeyCascadeDepth = 15
+
+func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, isPessimistic bool, depth int) error {
 	exec, ok := e.(WithForeignKeyTrigger)
 	if !ok {
 		return nil
 	}
+	originStmtCtx := a.Ctx.GetSessionVars().StmtCtx
+	defer func() {
+		// Reset to the original stmtCtx.
+		a.Ctx.GetSessionVars().StmtCtx = originStmtCtx
```

Comment:
I think `stmtctx` is not designed to be used like this. It's way too complex to reason every field in it in this case. I bet there are bugs here.

Reply:
You are right, I should handle `stmtctx` carefully.

Reply:
I have refactored the logic here, PTAL

Reply:
Does it mean 2 statements may share some of the same `stmtctx` now? I doubt that's still error-prone

Reply:
yes..., but the foreign key cascade SQL shouldn't be treated as a statement, it is another  `Executor`.

Reply:
I'm importing more tests from MySQL to test the correctness of this.
https://github.com/pingcap/tidb-test/pull/1993